alibi/explainers/similarity/grad.py

```diff
@@ -0,0 +1,203 @@
+from alibi.explainers.similarity.metrics import dot, cos, asym_dot
+from alibi.api.defaults import DEFAULT_META_SIM, DEFAULT_DATA_SIM
+
+if TYPE_CHECKING:
+    import tensorflow
+    import torch
+
+
+class SimilarityExplainer(BaseSimilarityExplainer):
+    def __init__(self,
+                 predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
+                 loss_fn: '''Callable[[Union[tensorflow.Tensor, torch.Tensor],
+                                    Union[tensorflow.Tensor, torch.Tensor]],
+                                   Union[tensorflow.Tensor, torch.Tensor]]''',
+                 sim_fn: 'Union[Callable[[np.ndarray, np.ndarray], np.ndarray], str]' = 'grad_dot',
+                 task: str = "classification",
+                 store_grads: bool = False,
+                 seed: int = 0,
+                 backend: str = "tensorflow",
+                 **kwargs
+                 ):
+        """Constructor
+
+        Parameters
+        ----------
+        predictor:
+            model to explain.
+        loss_fn:
+            Loss function used.
+        sim_fn:
+            Similarity function to use. ``'grad_dot'`` | ``'grad_cos'``. Default: ``'grad_dot'``.
+        task:
+            Task to perform. ``'classification'`` | ``'regression'``. Default: ``'classification'``.
+        store_grads:
+            Whether to store gradients. Default ``False``. If ``False``, gradients are computed on the fly otherwise we
+            store them which can be faster when it comes to explaining and instances.
+        seed:
+            Random seed. Default: 0.
+        backend:
+            Backend to use. ``'tensorflow'`` | ``'torch'``. Default: ``'tensorflow'``.
+        kwargs:
+            Additional arguments to pass to the similarity function.
+        """
+
+        self.meta = copy.deepcopy(DEFAULT_META_SIM)
+        self.meta['params'].update(
+            sim_fn_name=sim_fn,
+            store_grads=store_grads,
+            seed=seed,
+            backend_name=backend,
+            task_name=task
+        )
+
+        sim_fn_opts: Dict[str, Callable] = {
+            'grad_dot': dot,
+            'grad_cos': cos,
+            'grad_asym_dot': asym_dot
+        }
+
+        if isinstance(sim_fn, str):
+            if sim_fn not in sim_fn_opts.keys():
+                raise ValueError(f'Unknown method {sim_fn}. Consider using: `{"` | `".join(sim_fn_opts.keys())}`.')
+
+            sim_fn = sim_fn_opts[sim_fn]
+
+        if task not in ['classification', 'regression']:
+            raise ValueError(f'Unknown task {task}. Consider using: `classification` | `regression`.')
+
+        self.task = task
+
+        super().__init__(predictor, loss_fn, sim_fn, store_grads, seed, backend, meta=self.meta, **kwargs)
+
+    def _preprocess_args(
+            self,
+            X: 'Union[np.ndarray, tensorflow.Tensor, torch.Tensor]',
+            Y: 'Optional[Union[np.ndarray, tensorflow.Tensor, torch.Tensor, Callable]]' = None) \
```

Comment:
Why is it Y a Callable?
I think that is a bit confusing and I would prefer to stick with tensors/arrays

Reply:
I believe it's for cases when we can only get targets by calling the model (i.e. the Deploy use case)? It's similar for `IntegratedGradients` Would be good to hear more.

Reply:
@jklaise can you remind me what the use case was for IG? I'm struggling to recall why I did this, perhaps I was overthinking things and it's best to remove!

In the classification case, there's the default scenario where we use the maximum probability class. Is there any reason you might want to take a different class? In the regression case you have to supply the target explicitly (see [note](https://www.notion.so/Classification-Regression-Task-Issue-0f682117a9244863a3656e2ec9b42494)). Is there some reason you might want to compute a target dependent on the model output?

Reply:
The use case for IG is for Deploy reasons - you don't know which dimension (`argmax` of classification) you want to explain until you call the model, so for that reason we allow a "dynamic" selection based on a user `target_fn`. This is because in Deploy it's quite hard to pass in a model prediction into `explain` as things are not structured in a way to support it, and in essence it is not required as you can get the predictions internally in `explain` - just that in some cases like this the user needs to specify "dynamically" what to explain.

I believe the way it's implemented we don't have to support a "dynamic" `Callable` option - as you say, this is done automatically if `task='classification'` and `Y=None` and I don't think other use cases are that relevant.

Reply:
We could've done the same with `IntegratedGradients` but wanted to be more explicit (rather than doing `argmax` behind the scenes - especially since the current IG implementation has a hard time distinguishing classification and regression use cases which would complicate things even further, also the "dynamic" `target_fn` allows for multi-regression use cases).

Reply:
For reference have removed `Y` as a Callable.